File: scripts/release-rehearsal.py
#!/usr/bin/env python3
"""Fail-closed characterization of the exact-subject release rehearsal."""

from __future__ import annotations

import argparse
import hashlib
import json
import os
from pathlib import Path
import subprocess
import sys
from typing import Any, Callable

ROOT = Path(__file__).resolve().parent.parent
PHASE_COMPLETE = "Complete"
PHASE_INCOMPLETE = "Incomplete"
PHASE_MISMATCH = "Mismatch"
PHASE_INSTRUMENT_FAILURE = "InstrumentFailure"
CARGO_TOKEN_ENV = "CARGO_REGISTRY_TOKEN"


def compute_sha256(path: Path) -> str:
    """Return the repository's canonical SHA-256 text for one exact file."""
    digest = hashlib.sha256()
    with path.open("rb") as source:
        while chunk := source.read(65536):
            digest.update(chunk)
    return f"sha256:v1:{digest.hexdigest()}"


def resolve_commit(commit_ref: str) -> str:
    """Resolve one caller-supplied Git commit ref or fail without substitution."""
    if (
        not commit_ref
        or commit_ref.startswith("-")
        or any(char in commit_ref for char in "\r\n\0")
    ):
        raise ValueError("commit ref must be non-empty, single-line, and not start with a dash")
    result = subprocess.run(
        ["git", "rev-parse", "--verify", f"{commit_ref}^{{commit}}"],
        cwd=ROOT,
        capture_output=True,
        text=True,
        timeout=15,
        check=False,
    )
    if result.returncode != 0:
        raise ValueError(f"commit ref is not an exact repository commit: {commit_ref}")
    commit_sha = result.stdout.strip().lower()
    if len(commit_sha) not in (40, 64) or any(
        char not in "0123456789abcdef" for char in commit_sha
    ):
        raise ValueError("resolved commit identity is not canonical hexadecimal Git output")
    return commit_sha


def _file_characterization(path: Path) -> str:
    """Keep file presence explicit without upgrading it to semantic proof."""
    try:
        return PHASE_INCOMPLETE if path.is_file() else PHASE_MISMATCH
    except OSError:
        return PHASE_INSTRUMENT_FAILURE


def _sanitized_environment() -> dict[str, str]:
    """Prevent child characterizations from receiving the registry secret."""
    environment = dict(os.environ)
    environment.pop(CARGO_TOKEN_ENV, None)
    return environment


def _run_characterization(command: list[str]) -> str:
    """Run a bounded characterization without treating exit zero as Complete."""
    try:
        result = subprocess.run(
            command,
            cwd=ROOT,
            env=_sanitized_environment(),
            capture_output=True,
            text=True,
            timeout=300,
            check=False,
        )
    except (OSError, subprocess.SubprocessError):
        return PHASE_INSTRUMENT_FAILURE
    return PHASE_INCOMPLETE if result.returncode == 0 else PHASE_MISMATCH


def _workspace_version() -> str:
    """Read the exact workspace version. Source identity only: the grammar,
    tag, and channel decisions belong to the typed authority invoked next."""
    in_workspace_package = False
    for line in (ROOT / "Cargo.toml").read_text(encoding="utf-8").splitlines():
        stripped = line.strip()
        if stripped.startswith("[") and stripped.endswith("]"):
            in_workspace_package = stripped == "[workspace.package]"
            continue
        if in_workspace_package and stripped.startswith("version"):
            value = stripped.split("=", 1)[1].strip().strip('"')
            if value:
                return value
    raise ValueError("Cargo.toml has no [workspace.package] version")


def run_phase_release_identity(receipt: dict[str, Any]) -> str:
    """Consume the typed release-identity projection for the workspace candidate.

    The workspace version is read as the identity source and validated through
    ``cargo-allow release-identity``; this phase records the validated fields
    without re-deriving grammar, tag, or channel.
    """
    try:
        version = _workspace_version()
        result = subprocess.run(
            [
                "cargo", "run", "--quiet", "-p", "cargo-allow", "--locked", "--",
                "release-identity", "--version", version,
            ],
            cwd=ROOT,
            env=_sanitized_environment(),
            capture_output=True,
            text=True,
            timeout=300,
            check=False,
        )
    except (OSError, ValueError, subprocess.SubprocessError):
        return PHASE_INSTRUMENT_FAILURE
    if result.returncode != 0:
        return PHASE_MISMATCH
    try:
        projection = json.loads(result.stdout)
    except json.JSONDecodeError:
        return PHASE_INSTRUMENT_FAILURE
    if (
        projection.get("schema") != "cargo-allow.release-identity.v1"
        or projection.get("result") != "validated"
    ):
        return PHASE_MISMATCH
    receipt["release_identity"] = {
        "schema": projection["schema"],
        "version": projection["version"],
        "tag": projection["tag"],
        "tag_source": projection["tag_source"],
        "channel": projection["channel"],
        "rc_ordinal": projection["rc_ordinal"],
        "github_prerelease": projection["github_prerelease"],
    }
    return PHASE_COMPLETE


def run_phase_candidate_package_set(receipt: dict[str, Any]) -> str:
    """Package the ten cargo-allow candidate rows from the committed subject.

    The publisher's --package-only mode validates exact release-coupled
    internal requirements and packages the selected rows offline (no registry
    reads, no compilation); this phase records each row's name, version,
    release order, SHA-256, and archive size without any upload path.
    """
    preflight_path = ROOT / "target/cargo-allow/rehearsal-candidate-package-set.json"
    try:
        preflight_path.parent.mkdir(parents=True, exist_ok=True)
        result = subprocess.run(
            [
                sys.executable,
                (ROOT / "scripts/release-topology-publisher.py").as_posix(),
                "--mode", "cargo-allow",
                "--package-only",
                "--receipt", str(preflight_path),
            ],
            cwd=ROOT,
            env=_sanitized_environment(),
            capture_output=True,
            text=True,
            timeout=300,
            check=False,
        )
    except (OSError, subprocess.SubprocessError):
        return PHASE_INSTRUMENT_FAILURE
    rows: list[dict[str, Any]] = []
    try:
        packaged = json.loads(preflight_path.read_text(encoding="utf-8"))
        rows = packaged.get("rows", [])
    except (OSError, json.JSONDecodeError):
        if result.returncode == 0:
            return PHASE_INSTRUMENT_FAILURE
    recorded = []
    for row in rows:
        crate_path = Path(row["crate"])
        if not crate_path.is_absolute():
            crate_path = ROOT / crate_path
        try:
            size_bytes = crate_path.stat().st_size
        except OSError:
            size_bytes = None
        recorded.append(
            {
                "name": row.get("name"),
                "version": row.get("version"),
                "release_order": row.get("release_order"),
                "sha256": row.get("local_checksum"),
                "size_bytes": size_bytes,
            }
        )
    receipt["candidate_package_set"] = {"rows": recorded}
    if result.returncode != 0:
        return PHASE_MISMATCH
    expected_version = (receipt.get("release_identity") or {}).get("version")
    if (
        len(recorded) != 10
        or expected_version is None
        or any(row.get("version") != expected_version for row in recorded)
        or any(row.get("size_bytes") is None or row["size_bytes"] <= 0 for row in recorded)
    ):
        return PHASE_MISMATCH
    return PHASE_COMPLETE


def run_phase_shared_prerequisites(receipt: dict[str, Any]) -> str:
    """Prove the three topology-selected shared rows against retained
    namespace checksums through the read-only #3744 registry preflight.

    The preflight queries crates.io anonymously (the sanitized environment
    strips the publish token it must never see) and exits nonzero when any
    shared row is not already_published_exact; the recorded rows carry each
    row's observed state for diagnostics.
    """
    preflight_path = ROOT / "target/cargo-allow/rehearsal-shared-preflight.json"
    try:
        preflight_path.parent.mkdir(parents=True, exist_ok=True)
        result = subprocess.run(
            [
                sys.executable,
                (ROOT / "scripts/release-topology-publisher.py").as_posix(),
                "--mode", "cargo-allow",
                "--registry-preflight",
                "--receipt", str(preflight_path),
            ],
            cwd=ROOT,
            env=_sanitized_environment(),
            capture_output=True,
            text=True,
            timeout=300,
            check=False,
        )
    except (OSError, subprocess.SubprocessError):
        return PHASE_INSTRUMENT_FAILURE
    rows: list[dict[str, Any]] = []
    try:
        preflight = json.loads(preflight_path.read_text(encoding="utf-8"))
        rows = preflight.get("shared_registry_preflight", [])
    except (OSError, json.JSONDecodeError):
        if result.returncode == 0:
            return PHASE_INSTRUMENT_FAILURE
    receipt["shared_prerequisites"] = [
        {
            "name": row.get("name"),
            "version": row.get("version"),
            "state": row.get("state"),
            "registry_checksum": row.get("registry_checksum"),
        }
        for row in rows
    ]
    if result.returncode != 0:
        return PHASE_MISMATCH
    if not rows or len(rows) != 3 or any(
        row.get("state") != "already_published_exact" for row in rows
    ):
        return PHASE_MISMATCH
    return PHASE_COMPLETE


def _run_proof(command: list[str]) -> str:
    """Run a bounded fixture proof: exit zero proves, nonzero mismatches."""
    try:
        result = subprocess.run(
            command,
            cwd=ROOT,
            env=_sanitized_environment(),
            capture_output=True,
            text=True,
            timeout=300,
            check=False,
        )
    except (OSError, subprocess.SubprocessError):
        return PHASE_INSTRUMENT_FAILURE
    return PHASE_COMPLETE if result.returncode == 0 else PHASE_MISMATCH


def run_phase_publisher_state_machine(receipt: dict[str, Any]) -> str:
    """Exercise the publisher state machine fixture matrix (#3751 phase 4).

    The bounded fixture suite proves the upload state machine without any real
    publication: missing/existing-exact/conflicting rows, post-upload checksum
    conflicts, recovery exactness, malformed and unavailable provider
    responses, the shared prerequisite preflight matrix, and the
    shared-preflight-before-upload source law. No fixture calls real
    ``cargo publish`` or touches a token.
    """
    receipt["publisher_state_machine"] = {
        "fixture_matrix": "scripts/test-release-topology-publisher.py"
    }
    return _run_proof(
        [sys.executable, str(ROOT / "scripts/test-release-topology-publisher.py")]
    )


def run_phase_docs_and_support(receipt: dict[str, Any]) -> str:
    """Bind the typed identity to the docs and support surfaces (#3751 phase 5).

    Proves the changelog corpus reproduces CHANGELOG.md exactly, the typed
    identity's release record and GitHub note exist under docs/release/, and
    the support matrix and getting-started guide name the exact identity.
    The pinned-Changie merge roundtrip itself is proven continuously by the
    changie-contract CI lane; this phase proves the identity binding.
    """
    history_check = _run_proof(
        [sys.executable, str(ROOT / "scripts/generate-changie-history.py"), "--check"]
    )
    if history_check != PHASE_COMPLETE:
        return history_check

    identity = receipt.get("release_identity") or {}
    version = identity.get("version")
    tag = identity.get("tag")
    if not version or not tag:
        return PHASE_INCOMPLETE
    release_record = ROOT / "docs/release" / f"{version}.md"
    github_note = ROOT / "docs/release/github" / f"{tag}.md"
    support_matrix = ROOT / "docs/support-matrix.toml"
    getting_started = ROOT / "docs/getting-started.md"
    try:
        surfaces = {
            "release_record": release_record,
            "github_note": github_note,
            "support_matrix": support_matrix,
            "getting_started": getting_started,
        }
        contents = {
            name: path.read_text(encoding="utf-8") for name, path in surfaces.items()
        }
    except (OSError, UnicodeDecodeError):
        return PHASE_MISMATCH

    if not contents["support_matrix"].strip():
        return PHASE_MISMATCH
    if version not in contents["getting_started"]:
        return PHASE_MISMATCH

    receipt["docs_and_support_identity"] = {
        "release_record": release_record.as_posix(),
        "github_note": github_note.as_posix(),
        "support_matrix": support_matrix.as_posix(),
        "getting_started": getting_started.as_posix(),
        "history_check": "scripts/generate-changie-history.py --check",
    }
    return PHASE_COMPLETE


def run_phase_manifest_and_assets(receipt: dict[str, Any]) -> str:
    """Prove the manifest/asset surface tooling fixture matrix (#3751 phase 6).

    The bounded fixture suite proves final-packaged-surface reconciliation on
    actual crate bytes: archive digest and size binding, missing declared
    assets reconciling to Incomplete, per-crate identity via the CLI,
    prerelease and hyphenated identity preservation, and unexpected-archive
    rejection — all offline, with no real publication.
    """
    receipt["manifest_and_assets"] = {
        "fixture_matrix": "scripts/test-final-packaged-surface.py"
    }
    return _run_proof(
        [sys.executable, str(ROOT / "scripts/test-final-packaged-surface.py")]
    )


AUTHORIZATION_ARTIFACT = "release/authorize-v0.2.0.json"
AUTHORIZATION_SCHEMA = "cargo-allow.release-authorization.v1"


def run_phase_authorization_boundary(receipt: dict[str, Any]) -> str:
    """Prove the token-free rehearsal posture and bind the checked
    authorization artifact's identity (#3751 phase: authorization boundary).

    The rehearsal never consumes authorization: a publish token in the
    environment is an instrument failure, and the checked
    CargoAllowReleaseAuthorizationV1 artifact is read only to record which
    release identity would gate the authorized run. The phase deliberately
    stays Incomplete — only the authorized run (#3760/#2502) can claim
    authorization.
    """
    if os.environ.get(CARGO_TOKEN_ENV):
        return PHASE_INSTRUMENT_FAILURE
    try:
        artifact = json.loads(
            (ROOT / AUTHORIZATION_ARTIFACT).read_text(encoding="utf-8")
        )
    except (OSError, json.JSONDecodeError):
        return PHASE_MISMATCH
    release = artifact.get("release")
    commit = artifact.get("candidate_parent_commit")
    tree = artifact.get("candidate_parent_tree")
    lock = artifact.get("cargo_lock_sha256")
    hex_ok = lambda value: (
        isinstance(value, str)
        and len(value) in (40, 64)
        and all(char in "0123456789abcdef" for char in value)
    )
    # The checked artifact stores the lock digest in its own bare-hex form
    # (no sha256: prefix); the phase validates that stored form rather than
    # imposing the receipt convention or re-deriving the value.
    lock_ok = lambda value: (
        isinstance(value, str)
        and len(value) == 64
        and all(char in "0123456789abcdef" for char in value)
    )
    if (
        artifact.get("schema_id") != AUTHORIZATION_SCHEMA
        or not isinstance(release, str)
        or not release.startswith("v")
        or not hex_ok(commit)
        or not hex_ok(tree)
        or not lock_ok(lock)
    ):
        return PHASE_MISMATCH
    receipt["authorization_boundary"] = {
        "authorization_artifact": AUTHORIZATION_ARTIFACT,
        "schema": AUTHORIZATION_SCHEMA,
        "named_release": release,
        "candidate_commit": commit,
        "token_present": False,
        "phase_status_note": (
            "deliberately Incomplete: the rehearsal never consumes "
            "authorization; #3760/#2502 gate the authorized run"
        ),
    }
    return PHASE_INCOMPLETE


def _yaml_workflow_inventory(path: Path) -> dict[str, Any] | None:
    """Parse one workflow manifest with PyYAML; None when unavailable."""
    try:
        import yaml
    except ImportError:
        return None
    try:
        parsed = yaml.safe_load(path.read_text(encoding="utf-8"))
    except Exception as exc:
        return {"error": f"YAML parse error: {exc}"}
    if not isinstance(parsed, dict) or "jobs" not in parsed:
        return {"error": "workflow is not a mapping with jobs"}
    jobs = {
        job_name: {
            "permissions": job.get("permissions"),
            "runs_on": job.get("runs-on"),
        }
        for job_name, job in parsed["jobs"].items()
        if isinstance(job, dict)
    }
    return {"top_level_permissions": parsed.get("permissions"), "jobs": jobs}


def run_phase_workflow_graph_permissions(receipt: dict[str, Any]) -> str:
    """Bind the release workflow graph's permission surface (#3751 phase).

    Enforces the least-privilege law — top-level ``actions: read`` and
    ``contents: write``, with ``github-release`` as the write/OIDC-scoped
    job and the authorized namespace workflow in namespace mode — and
    records the proof. PyYAML is used when available; otherwise the same
    law is checked on the pinned manifest strings.
    """
    release_path = ROOT / ".github/workflows/release.yml"
    authorized_path = ROOT / ".github/workflows/release-authorized.yml"
    try:
        release_text = release_path.read_text(encoding="utf-8")
        authorized_text = authorized_path.read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError):
        return PHASE_MISMATCH

    parsed = _yaml_workflow_inventory(release_path)
    if parsed is not None and "error" in parsed:
        return PHASE_MISMATCH

    authorized_namespace_mode = "--mode namespace" in authorized_text
    if parsed is None:
        proof = {
            "mode": "text",
            "release_jobs": [],
            "privileged_jobs": [],
            "top_level_read_scoped": "actions: read" in release_text,
            "top_level_write_scoped": "contents: write" in release_text,
            "github_release_scoped": "id-token: write" in release_text,
            "authorized_namespace_mode": authorized_namespace_mode,
        }
    else:
        jobs = parsed["jobs"]
        if not isinstance(jobs, dict) or not jobs:
            return PHASE_MISMATCH
        privileged = sorted(
            job_name
            for job_name, job in jobs.items()
            if isinstance(job.get("permissions"), dict)
        )
        proof = {
            "mode": "yaml",
            "release_jobs": sorted(jobs),
            "privileged_jobs": privileged,
            "top_level_read_scoped": parsed.get("top_level_permissions", {}).get(
                "actions"
            )
            == "read",
            "top_level_write_scoped": parsed.get("top_level_permissions", {}).get(
                "contents"
            )
            == "write",
            "github_release_scoped": jobs.get("github-release", {}).get(
                "permissions", {}
            ).get("id-token")
            == "write",
            "authorized_namespace_mode": authorized_namespace_mode,
        }

    if not (
        proof["top_level_read_scoped"]
        and proof["top_level_write_scoped"]
        and proof["github_release_scoped"]
        and proof["authorized_namespace_mode"]
    ):
        return PHASE_MISMATCH
    receipt["workflow_graph_permissions"] = proof
    return PHASE_COMPLETE


CHARACTERIZATION_PHASES = frozenset({
    "authorization_boundary",
})


def _aggregate_phase_status(phases: dict[str, str]) -> str:
    """Fail-closed aggregate: Complete only when every real phase proves and no
    characterization-only phase can manufacture that status."""
    values = set(phases.values())
    if PHASE_INSTRUMENT_FAILURE in values:
        return PHASE_INSTRUMENT_FAILURE
    if PHASE_MISMATCH in values:
        return PHASE_MISMATCH
    if any(phases.get(name) == PHASE_COMPLETE for name in CHARACTERIZATION_PHASES):
        return PHASE_MISMATCH
    return PHASE_INCOMPLETE


def _write_receipt(path: Path, json_text: str) -> None:
    """Write a receipt without following a symlink at the output leaf."""
    if path.is_symlink() or path.is_dir():
        raise OSError("output path cannot be a symlink or directory")
    path.parent.mkdir(parents=True, exist_ok=True)
    if path.is_symlink() or path.is_dir():
        raise OSError("output path cannot be a symlink or directory")

    flags = os.O_WRONLY | os.O_CREAT | os.O_TRUNC
    flags |= getattr(os, "O_NOFOLLOW", 0)
    descriptor = os.open(path, flags, 0o600)
    try:
        with os.fdopen(descriptor, "w", encoding="utf-8") as output:
            descriptor = -1
            output.write(json_text + "\n")
    finally:
        if descriptor != -1:
            os.close(descriptor)


def build_rehearsal_receipt(commit_ref: str) -> dict[str, Any]:
    """Build an honest characterization receipt for one verified commit."""
    commit_sha = resolve_commit(commit_ref)
    lockfile_digest = compute_sha256(ROOT / "Cargo.lock")
    topology_digest = compute_sha256(
        ROOT / "policy/product-package-topology-v2.toml"
    )

    receipt: dict[str, Any] = {
        "schema_version": "1.0",
        "receipt_id": f"REHEARSAL-{commit_sha[:8]}",
        "commit_sha": commit_sha,
        "subject_lockfile_digest": lockfile_digest,
        "subject_topology_digest": topology_digest,
        "zero_mutation_proof": {
            "tag_mutation_prevented": False,
            "token_read_prevented": False,
            "cargo_publish_prevented": False,
            "registry_mutation_prevented": False,
            "github_release_mutation_prevented": False,
            "live_setting_mutation_prevented": False,
            "external_repository_mutation_prevented": False,
        },
        "phases": {},
        "aggregate_status": PHASE_INCOMPLETE,
        "claim_boundary": (
            "Phases release_identity, candidate_package_set, "
            "shared_prerequisites, publisher_state_machine, "
            "docs_and_support_identity, and manifest_and_assets prove typed "
            "semantics (typed identity validation; offline candidate packaging "
            "with exact internal requirements; read-only shared registry "
            "equality; the publisher fixture state-machine matrix; "
            "changelog-corpus identity plus exact release-record/note and "
            "support-doc binding; the manifest/asset surface fixture matrix; "
            "the release workflow graph permission inventory). The "
            "authorization_boundary phase deliberately stays Incomplete: the "
            "rehearsal proves the token-free posture and records the checked "
            "authorization artifact's identity but never consumes "
            "authorization, so the aggregate cannot satisfy a release gate."
        ),
    }

    phases: dict[str, Callable[[dict[str, Any]], str]] = {
        "release_identity": run_phase_release_identity,
        "candidate_package_set": run_phase_candidate_package_set,
        "shared_prerequisites": run_phase_shared_prerequisites,
        "publisher_state_machine": run_phase_publisher_state_machine,
        "docs_and_support_identity": run_phase_docs_and_support,
        "manifest_and_assets": run_phase_manifest_and_assets,
        "authorization_boundary": run_phase_authorization_boundary,
        "workflow_graph_permissions": run_phase_workflow_graph_permissions,
    }

    for phase_name, runner in phases.items():
        receipt["phases"][phase_name] = runner(receipt)

    receipt["aggregate_status"] = _aggregate_phase_status(receipt["phases"])
    return receipt


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Characterize the exact-subject release rehearsal fail-closed"
    )
    parser.add_argument("--commit", default="HEAD", help="Exact Git commit or ref")
    parser.add_argument("--output", help="Path to write receipt JSON")
    args = parser.parse_args()

    try:
        receipt = build_rehearsal_receipt(args.commit)
    except (OSError, ValueError, subprocess.SubprocessError) as error:
        print(f"release rehearsal instrumentation failed: {error}", file=sys.stderr)
        return 2

    json_text = json.dumps(receipt, indent=2, sort_keys=True)
    if args.output:
        output_path = Path(args.output)
        _write_receipt(output_path, json_text)
        print(f"Receipt written to {output_path}")
    else:
        print(json_text)

    return 0 if receipt["aggregate_status"] == PHASE_COMPLETE else 1


if __name__ == "__main__":
    sys.exit(main())

File: crates/cargo-allow/tests/release_rehearsal.rs
use std::error::Error;
use std::io;
use std::path::{Path, PathBuf};
use std::process::Command;

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
struct ZeroMutationProof {
    tag_mutation_prevented: bool,
    token_read_prevented: bool,
    cargo_publish_prevented: bool,
    registry_mutation_prevented: bool,
    github_release_mutation_prevented: bool,
    live_setting_mutation_prevented: bool,
    external_repository_mutation_prevented: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
struct ReleaseRehearsalReceiptV1 {
    schema_version: String,
    receipt_id: String,
    commit_sha: String,
    subject_lockfile_digest: String,
    subject_topology_digest: String,
    zero_mutation_proof: ZeroMutationProof,
    phases: std::collections::BTreeMap<String, String>,
    #[serde(default)]
    release_identity: Option<ReleaseIdentityRecordV1>,
    #[serde(default)]
    shared_prerequisites: Option<Vec<SharedPrerequisiteRowV1>>,
    #[serde(default)]
    candidate_package_set: Option<CandidatePackageSetRecordV1>,
    #[serde(default)]
    publisher_state_machine: Option<PublisherStateMachineRecordV1>,
    #[serde(default)]
    docs_and_support_identity: Option<DocsAndSupportIdentityRecordV1>,
    #[serde(default)]
    manifest_and_assets: Option<ManifestAndAssetsRecordV1>,
    #[serde(default)]
    workflow_graph_permissions: Option<WorkflowGraphPermissionsRecordV1>,
    #[serde(default)]
    authorization_boundary: Option<AuthorizationBoundaryRecordV1>,
    aggregate_status: String,
    claim_boundary: String,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct WorkflowGraphPermissionsRecordV1 {
    mode: String,
    release_jobs: Vec<String>,
    privileged_jobs: Vec<String>,
    top_level_read_scoped: bool,
    top_level_write_scoped: bool,
    github_release_scoped: bool,
    authorized_namespace_mode: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct AuthorizationBoundaryRecordV1 {
    authorization_artifact: String,
    schema: String,
    named_release: String,
    candidate_commit: String,
    token_present: bool,
    phase_status_note: String,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct ManifestAndAssetsRecordV1 {
    fixture_matrix: String,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct DocsAndSupportIdentityRecordV1 {
    release_record: String,
    github_note: String,
    support_matrix: String,
    getting_started: String,
    history_check: String,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct PublisherStateMachineRecordV1 {
    fixture_matrix: String,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct CandidatePackageSetRecordV1 {
    rows: Vec<CandidatePackageRowV1>,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct CandidatePackageRowV1 {
    name: String,
    version: String,
    release_order: u32,
    sha256: String,
    size_bytes: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct SharedPrerequisiteRowV1 {
    name: String,
    version: String,
    state: String,
    registry_checksum: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct ReleaseIdentityRecordV1 {
    schema: String,
    version: String,
    tag: String,
    tag_source: String,
    channel: String,
    rc_ordinal: Option<u32>,
    github_prerelease: bool,
}

fn require(condition: bool, message: &str) -> Result<(), io::Error> {
    if condition {
        Ok(())
    } else {
        Err(io::Error::other(message))
    }
}

fn repo_root() -> Result<PathBuf, Box<dyn Error>> {
    let manifest_dir = Path::new(env!("CARGO_MANIFEST_DIR"));
    let crates_dir = manifest_dir
        .parent()
        .ok_or_else(|| io::Error::other("no crates dir parent"))?;
    let root = crates_dir
        .parent()
        .ok_or_else(|| io::Error::other("no repo root"))?;
    Ok(root.to_path_buf())
}

#[test]
fn rehearsal_characterization_fails_closed() -> Result<(), Box<dyn Error>> {
    let root = repo_root()?;
    let script = root.join("scripts/release-rehearsal.py");
    require(script.is_file(), "release rehearsal script is missing")?;

    let output = Command::new("python")
        .arg(&script)
        .arg("--commit")
        .arg("HEAD")
        .current_dir(&root)
        .output()?;

    require(
        output.status.code() == Some(1),
        &format!(
            "characterization must exit one, got {:?}: {}",
            output.status.code(),
            String::from_utf8_lossy(&output.stderr)
        ),
    )?;

    let receipt: ReleaseRehearsalReceiptV1 = serde_json::from_slice(&output.stdout)?;
    require(
        receipt.schema_version == "1.0",
        "schema version must be 1.0",
    )?;
    require(
        receipt.receipt_id.starts_with("REHEARSAL-"),
        "receipt ID must name the resolved commit",
    )?;
    require(
        receipt.aggregate_status != "Complete",
        "characterization must not report Complete",
    )?;
    require(
        receipt
            .claim_boundary
            .contains("cannot satisfy a release gate"),
        "claim boundary must retain the characterization limitation",
    )?;
    require(
        receipt.subject_lockfile_digest.starts_with("sha256:v1:"),
        "lockfile digest must use canonical SHA-256 text",
    )?;
    require(
        receipt.subject_topology_digest.starts_with("sha256:v1:"),
        "topology digest must use canonical SHA-256 text",
    )?;
    require(
        matches!(receipt.commit_sha.len(), 40 | 64)
            && receipt
                .commit_sha
                .bytes()
                .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte)),
        "commit identity must be canonical lowercase hexadecimal",
    )?;

    let proof = &receipt.zero_mutation_proof;
    require(
        [
            proof.tag_mutation_prevented,
            proof.token_read_prevented,
            proof.cargo_publish_prevented,
            proof.registry_mutation_prevented,
            proof.github_release_mutation_prevented,
            proof.live_setting_mutation_prevented,
            proof.external_repository_mutation_prevented,
        ]
        .into_iter()
        .all(|value| !value),
        "unproven zero-mutation facts must remain false",
    )?;

    // The two characterization-only phases can never manufacture
    // completion; release_identity through manifest_and_assets are real
    // phases (#3751 phases 1-6) and may report Complete when their proofs
    // succeed.
    // The authorization_boundary phase deliberately stays Incomplete: the
    // rehearsal never consumes authorization (#3760/#2502 gate the real
    // run). workflow_graph_permissions is a real phase (#3751 phase 7) and
    // may report Complete when its proof succeeds.
    require(
        receipt
            .phases
            .get("authorization_boundary")
            .is_some_and(|status| status != "Complete"),
        "phase authorization_boundary must exist and remain non-Complete",
    )?;
    require(
        receipt
            .phases
            .get("release_identity")
            .is_some_and(|status| !status.is_empty()),
        "the typed release_identity phase must report a status",
    )?;
    let identity = receipt.release_identity.as_ref().ok_or_else(|| {
        io::Error::other("a validated release_identity phase must record the typed projection")
    })?;
    require(
        identity.schema == "cargo-allow.release-identity.v1",
        "the recorded identity must carry the typed schema identity",
    )?;
    require(
        !identity.version.is_empty(),
        "identity version must be recorded",
    )?;
    require(
        identity.tag.starts_with('v'),
        "the canonical tag must be recorded",
    )?;
    require(
        (identity.channel == "stable") != (identity.channel == "release_candidate"),
        "channel must be exactly one of stable or release_candidate",
    )?;
    require(
        identity.github_prerelease == (identity.channel == "release_candidate"),
        "GitHub prerelease posture must follow the channel",
    )?;

    let packages = receipt.candidate_package_set.as_ref().ok_or_else(|| {
        io::Error::other("a validated candidate_package_set phase must record the packaged rows")
    })?;
    require(
        packages.rows.len() == 10,
        "the candidate set must package exactly ten rows",
    )?;
    let identity_version = &identity.version;
    for row in &packages.rows {
        require(
            &row.version == identity_version,
            "every packaged row must carry the selected release identity version",
        )?;
        require(
            row.sha256.starts_with("sha256:"),
            "a packaged row must record a canonical sha256 digest",
        )?;
        require(
            row.size_bytes > 0,
            "a packaged row must record a positive size",
        )?;
    }

    let machine_status = receipt
        .phases
        .get("publisher_state_machine")
        .ok_or_else(|| io::Error::other("publisher_state_machine phase must exist"))?;
    require(
        machine_status == "Complete",
        "the offline publisher state-machine fixture matrix must prove Complete",
    )?;
    let machine = receipt.publisher_state_machine.as_ref().ok_or_else(|| {
        io::Error::other("a proven publisher_state_machine phase must record its fixture matrix")
    })?;
    require(
        machine.fixture_matrix == "scripts/test-release-topology-publisher.py",
        "the fixture matrix provenance must name the publisher contract suite",
    )?;

    let assets_status = receipt
        .phases
        .get("manifest_and_assets")
        .ok_or_else(|| io::Error::other("manifest_and_assets phase must exist"))?;
    require(
        assets_status == "Complete",
        "the offline manifest/asset fixture matrix must prove Complete",
    )?;
    let assets = receipt.manifest_and_assets.as_ref().ok_or_else(|| {
        io::Error::other("a proven manifest_and_assets phase must record its fixture matrix")
    })?;
    require(
        assets.fixture_matrix == "scripts/test-final-packaged-surface.py",
        "the fixture matrix provenance must name the surface contract suite",
    )?;

    let workflow_status = receipt
        .phases
        .get("workflow_graph_permissions")
        .ok_or_else(|| io::Error::other("workflow_graph_permissions phase must exist"))?;
    require(
        workflow_status == "Complete",
        "the workflow graph permission inventory must prove Complete",
    )?;
    let workflow = receipt.workflow_graph_permissions.as_ref().ok_or_else(|| {
        io::Error::other("a proven workflow_graph_permissions phase must record its inventory")
    })?;
    require(
        workflow.top_level_read_scoped
            && workflow.top_level_write_scoped
            && workflow.github_release_scoped
            && workflow.authorized_namespace_mode,
        "the recorded workflow graph proof must carry every least-privilege law",
    )?;
    let authorization = receipt.authorization_boundary.as_ref().ok_or_else(|| {
        io::Error::other("the authorization boundary phase must record the checked artifact")
    })?;
    require(
        !authorization.token_present,
        "the rehearsal must prove the publish token was absent",
    )?;
    require(
        authorization.named_release.starts_with('v'),
        "the checked authorization artifact must name its release",
    )?;

    let docs_status = receipt
        .phases
        .get("docs_and_support_identity")
        .ok_or_else(|| io::Error::other("docs_and_support_identity phase must exist"))?;
    require(
        docs_status == "Complete",
        "the docs/support identity binding must prove Complete",
    )?;
    let docs = receipt.docs_and_support_identity.as_ref().ok_or_else(|| {
        io::Error::other("a proven docs_and_support_identity phase must record its surfaces")
    })?;
    require(
        docs.release_record
            .ends_with(&format!("/{}.md", identity.version)),
        "the release record must be bound to the typed identity version",
    )?;
    require(
        docs.github_note
            .ends_with(&format!("/github/{}.md", identity.tag)),
        "the GitHub note must be bound to the typed identity tag",
    )?;

    let shared = receipt.shared_prerequisites.as_ref().ok_or_else(|| {
        io::Error::other("a validated shared_prerequisites phase must record the preflight rows")
    })?;
    require(
        shared.len() == 3,
        "the shared preflight must record exactly three rows",
    )?;
    for row in shared {
        require(
            row.state == "already_published_exact",
            "every shared prerequisite must be already_published_exact for the phase to prove",
        )?;
        require(
            row.registry_checksum
                .as_deref()
                .is_some_and(|checksum| checksum.starts_with("sha256:")),
            "an exact shared row must record its canonical registry checksum",
        )?;
    }

    Ok(())
}
